### PR TITLE
Reorganized template

### PR DESCRIPTION
**Pull request by ancestor-mithril**
* Refactored and decoupled some methods from main.
* Reorganized datasets: 
   * now cached transformations are done at the beginning of training.
* Reorganized metrics:
  * metrics are now simpler;
  * they are more flexible and could use any information passed in kwargs.
* Moved stuff from training in separate functions, which can be more easily commented.

### CODE REVIEW

File: configs/config.yaml [hunk 1]
```diff
@@ -1,3 +1,7 @@
+hydra:
+  job:
+    chdir: True
```
Comment (ancestor-mithril) [inline configs/config.yaml:3]: This is what need to be added to suppress hydra-core warnings.

Comment (simi2525) [reply configs/config.yaml:3]: Which warnings? If they are related to deprecated features, we should probably upgrade to the latest and fix the differences

Comment (ancestor-mithril) [reply configs/config.yaml:3]: The warnings are related to changing the default directory of the job. The default behavior of new versions of hydra-core is that the default directory does not change (so everything will be saved to "./runs/...", not to "./outputs/runs/..."

**Review by simi2525 — APPROVED** [review]
LGTM, a good refresh in the main file, a good starting point to finish the TODOs left in the project, can expand when needed by future projects.

**Pull request merged**